man/iperender.1
.\"                                      Hey, EMACS: -*- nroff -*-
.\" First parameter, NAME, should be all caps
.\" Second parameter, SECTION, should be 1-8, maybe w/ subsection
.\" other parameters are allowed: see man(7), man(1)
.\" TeX users may be more comfortable with the \fB<whatever>\fP and
.\" \fI<whatever>\fP escape sequences to invode bold face and italics, 
.\" respectively.
.TH IPERENDER 1 "June 27, 2009"
.\" Please adjust this date whenever revising the manpage.
.\"
.\" Some roff macros, for reference:
.\" .nh        disable hyphenation
.\" .hy        enable hyphenation
.\" .ad l      left justify
.\" .ad b      justify to both left and right margins
.\" .nf        disable filling
.\" .fi        enable filling
.\" .br        insert line break
.\" .sp <n>    insert n+1 empty lines
.\" for manpage-specific macros, see man(7)
.SH NAME
iperender \- Export Ipe document as bitmap or SVG 
.SH SYNOPSIS
.B iperender
( -png | -eps | -svg ) 
[ -page \fIpage\fP ] 
[ -view \fIview\fP ]
[ -resolution \fIdpi\fP ]
[ -transparent ]
[ -nocrop ] 
\fIinput-file\fP \fIoutput-file\fP

.SH DESCRIPTION
.PP
\fBiperender\fP exports an Ipe document as a PNG bitmap, or as a
figure in Encapsulated Postscript (EPS) or in SVG format.
Options are:
.TP
\fB-png\fP
export as a PNG image
.TP
\fB-eps\fP
export as an EPS figure
.TP
\fB-svg\fP
export as an SVG figure
.TP
\fB-page\fP \fIpage\fP
export this page from a multipage document.
.TP
\fB-view\fP \fIview\fP
export this view from a page with multiple views.
.TP
\fB-resolution\fP \fIdpi\fP
set resolution for exporting to PNG.
.TP
\fB-transparent\fP
make background transparent when exporting to PNG.
.TP
\fB-nocrop\fP
do not crop the page to the bounding box of the objects.

.SH ENVIRONMENT VARIABLES

\fBiperender\fP respects the following environment variables:

.TP
\fBIPELATEXDIR\fP
the directory where \fBipetoipe\fP runs Pdflatex.
.TP
\fBIPELATEXPATH\fP
the directory containing pdflatex and xelatex.  If not set, they are
searched for on the path.
.TP
\fBIPEDEBUG\fP
set this to 1 for debugging output from \fBiperender\fP.

.SH AUTHOR
Otfried Cheong

.SH REPORTING BUGS
.ad l
Please report bugs using Ipe bugzilla at
.I "http://ipe.otfried.org/bugzilla.html"

.SH SEE ALSO
.ad l
Ipe is documented fully in
.IR "The Ipe Manual" ,
which is available online at
.I "http://ipe.otfried.org/manual/manual.html" .


news.txt
----------------------------------------------------------------------

Ipe 7.2.5 (2016-07-19) Bayreuth edition

 * The PDF interpreter for text objects has been improved to handle a
   much larger subset of PDF.  Using interesting PDF-generating latex
   packages is now supported and encouraged, including tikz, pgf, and
   tcolorbox.  This makes it easy to create boxes for theorems and the
   like in presentations, as the beamer package does.

 * The use of \includegraphics inside text objects is now supported.
   Since PDF is a very complex standard, some PDF features will not be
   displayed correctly on the screen (but will work correctly in Ipe's
   PDF output).  A few features cannot even be supported in the PDF
   output. 

 * luatex is now supported as an alternative to pdftex and xetex.  

 * New shear transformation mode.  Without axis set, it will use a
   horizontal line as the fixed line, otherwise your x-axis.

 * Groups can now be decorated using "decoration symbols".  Those
   should be a path (or a group of paths), and are automatically
   resized to fit the bounding box of the group.  You'll need to
   include a style sheet containing decoration symbols
   (e.g. "decorations.isy" from the distribution), then right click on
   a group to set a decoration.
 
 * Action "Edit object" can now be used to edit text inside a group
   object.  This is convenient, for instance, to edit graph vertices
   (which are most easily represented as a group consisting of a
   circle and a text object, or a mark and a text object).

 * Experimental recursive group edit implemented.  It keeps all state
   inside the Ipe document, so that you can save in the middle of a
   group edit (and autosave continues to work during the group edit).
   Group edit creates a new layer "EDIT GROUP", unpacks the group into
   this layer, and locks all other layers. "End group edit"
   automatically groups the elements again and returns the group to
   its original layer.

 * New action "set tangent direction" to set the axis to be tangent to
   a given circle (or ellipse) or arc.  Move mouse close to boundary
   of circle or circular arc and press Ctrl+F3.

 * Improvements for ink mode to make it more responsive, less
   memory-consuming, and display smoother strokes (thanks to Zhengdao
   Wang and Philipp Kindermann for contributing code).

 * Fixed missing shortcut for box mode.

 * Added "Apply properties" to object menus. Fixed bug that caused it
   to be applied to all objects (not just the selected ones).

 * When creating a path object, the current arrow size and shape is
   set, so that they can be used if an arrow is enabled later.

 * Snapping no longer occurs for objects in layers that are not
   visible (bug #88).

 * Text objects now keep their PDF representation as a separate
   XObject.  This implies that after changing the color of a text
   object, Latex needs to be run to see the change on screen. This
   should be automatic when automatic Latex run is enabled.

 * \ipesymbol is no longer supported - you can use tikz or pgf
   instead, or simply \includegraphics.

----------------------------------------------------------------------

Ipe 7.2.4 (2016-06-23) Bayreuth edition

 * New modes for making rectangles and parallelograms.

 * Page numbering works again (bug #57).  They are now customizable,
   see the manual.

 * Can now create links to external documents or websites, by
   right-clicking on any top-level group object and using "Set link
   URL" (feature #41).

 * Actions to pick the properties of an object (and set them in the
   properties panel), and to apply all current properties to the
   selected objects.

 * Replaced NSComboBox by NSPopUpButton.  They are more comfortable to
   use (you can click anywhere to change the setting, rather than just
   on the arrow), can display the color icons, and allow to re-select the
   currently selected choice (bug #47).
   
 * File format has been updated to 70204, but only if one of the two
   new features is used (external links, non-default pagenumberstyle).

 * Implemented double-clicks on OSX (bug #79).

 * Some support for HiDPI screens in Windows and Qt (use
   prefs.ui_scale and prefs.toolbar_scale).

 * Tested and documented installing Ipe with TeXlive on a USB-stick,
   so that it can be used on any Windows computer without installing
   anything (see manual).

 * Fixed combination of prefs.auto_external_editor and
   prefs.editor_closes_dialog.

 * Removed beginnings of magnification gesture handling on OSX (it
   wasn't finished and broke two-finger gestures for right-click).

 * Fixed Save dialog (bug #78).

 * Fixed behaviour when prefs.terminate_on_close is false on OSX (bug
   #51).

 * Improved width of grid size/angle size selector (bug #48).

 * Obey prefs.editable_textfile for editing style sheets (bug #80).

 * Updated ipe.dtd and manual for the correct Ipe file format (bug
   #67).

 * Customizable order of toolbars on Windows (feature #28).  

 * Fixed angle size menu on Qt.

 * Fixed join path for more than two segments (bug #85).

----------------------------------------------------------------------

Ipe 7.2.3 (2016-06-11) Bayreuth edition

 * Added /opt/local/bin to latex search path on OS X.

 * mainmenu.xib was missing from source distribution (bug #44).

 * Enabled "Command" modifier for mouse actions (bug #46).

 * Using Cairo FAST mode for rendering bitmaps (feature #53).

 * Added /bin and /usr/bin to PATH on OSX (bug #71).

 * Fixed crash in Ipelet information when ipelets contains no 'about'
   line (bug #66).

 * Fixed crash when processing empty text object (bug #64).

 * Notes and bookmarks can now be shown on startup on Cocoa by setting
   prefs (bug #68).

 * Fixed code so that compilation on older OSX versions is easier
   (feature #60).

 * Fixed substitution of quotes by typographic quotes on OSX (bug #69).

 * Removed duplicate overwrite warning on OSX (bug #50).


src/iperender/iperender.cpp
// --------------------------------------------------------------------
// iperender
// --------------------------------------------------------------------
/*

    This file is part of the extensible drawing editor Ipe.
    Copyright (C) 1993-2016  Otfried Cheong

    Ipe is free software; you can redistribute it and/or modify it
    under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 3 of the License, or
    (at your option) any later version.

    As a special exception, you have permission to link Ipe with the
    CGAL library and distribute executables, as long as you follow the
    requirements of the Gnu General Public License in regard to all of
    the software in the executable aside from CGAL.

    Ipe is distributed in the hope that it will be useful, but WITHOUT
    ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
    or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public
    License for more details.

    You should have received a copy of the GNU General Public License
    along with Ipe; if not, you can find it at
    "http://www.gnu.org/copyleft/gpl.html", or write to the Free
    Software Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

*/

#include "ipedoc.h"
#include "ipethumbs.h"

#include <cstdio>
#include <cstdlib>

using ipe::Document;
using ipe::Page;
using ipe::Thumbnail;

// --------------------------------------------------------------------

static int renderPage(Thumbnail::TargetFormat fm,
		      const char *src, const char *dst,
		      int pageNum, int viewNum, double zoom,
		      bool transparent, bool nocrop)
{
  Document *doc = Document::loadWithErrorReport(src);

  if (!doc)
    return 1;

  if (pageNum < 1 || pageNum > doc->countPages()) {
    fprintf(stderr,
	    "The document contains %d pages, cannot convert page %d.\n",
	    doc->countPages(), pageNum);
    delete doc;
    return 1;
  }

  if (doc->runLatex()) {
    delete doc;
    return 1;
  }

  Thumbnail tn(doc, 0);
  const Page *page = doc->page(pageNum - 1);
  if (!tn.saveRender(fm, dst, page, viewNum - 1, zoom, transparent, nocrop))
    fprintf(stderr, "Failure to render page.\n");
  delete doc;
  return 0;
}

// --------------------------------------------------------------------

static void usage()
{
  fprintf(stderr, "Usage: iperender [ -png ");
#ifdef CAIRO_HAS_PS_SURFACE
  fprintf(stderr, "| -eps ");
#endif
#ifdef CAIRO_HAS_PDF_SURFACE
  fprintf(stderr, "| -pdf ");
#endif
#ifdef CAIRO_HAS_SVG_SURFACE
  fprintf(stderr, "| -svg ");
#endif
  fprintf(stderr, "] "
	  "[ -page <page> ] [ -view <view> ] [ -resolution <dpi> ] "
	  "infile outfile\n"
	  "Iperender saves a single page of the Ipe document in some formats.\n"
	  " -page       : page to save (default 1).\n"
	  " -view       : view to save (default 1).\n"
	  " -resolution : resolution for png format (default 72.0 ppi).\n"
	  " -transparent: use transparent background in png format.\n"
	  " -nocrop     : do not crop page.\n"
	  );
  exit(1);
}

int main(int argc, char *argv[])
{
  ipe::Platform::initLib(ipe::IPELIB_VERSION);

  // ensure at least three arguments (handles -help as well :-)
  if (argc < 4)
    usage();

  Thumbnail::TargetFormat fm = Thumbnail::EPNG;
  if (!strcmp(argv[1], "-png"))
    fm = Thumbnail::EPNG;
#ifdef CAIRO_HAS_PS_SURFACE
  else if (!strcmp(argv[1], "-eps"))
    fm = Thumbnail::EPS;
#endif
#ifdef CAIRO_HAS_PDF_SURFACE
  else if (!strcmp(argv[1], "-pdf"))
    fm = Thumbnail::EPDF;
#endif
#ifdef CAIRO_HAS_SVG_SURFACE
  else if (!strcmp(argv[1], "-svg"))
    fm = Thumbnail::ESVG;
#endif
  else
    usage();

  int page = 1;
  int view = 1;
  double dpi = 72.0;
  int i = 2;
  bool transparent = false;
  bool nocrop = false;

  if (!strcmp(argv[i], "-page")) {
    page = ipe::Lex(ipe::String(argv[i+1])).getInt();
    i += 2;
  }

  if (!strcmp(argv[i], "-view")) {
    view = ipe::Lex(ipe::String(argv[i+1])).getInt();
    i += 2;
  }

  if (!strcmp(argv[i], "-resolution")) {
    dpi = ipe::Lex(ipe::String(argv[i+1])).getDouble();
    i += 2;
  }

  if (!strcmp(argv[i], "-transparent")) {
    transparent = true;
    ++i;
  }

  if (!strcmp(argv[i], "-nocrop")) {
    nocrop = true;
    ++i;
  }

  // remaining arguments must be two filenames
  if (argc != i + 2)
    usage();

  const char *src = argv[i];
  const char *dst = argv[i+1];

  return renderPage(fm, src, dst, page, view, dpi / 72.0,
		    transparent, nocrop);
}

// --------------------------------------------------------------------
